### DISCUSSION TITLE
How to backup and restore?

### DISCUSSION THREAD

**Discussion by zero-thermo** [wger-project/wger#1285] [Q&A]
What's the right way to backup and restore wger that is setup via docker-compose?  

I tried the following:
1) I used [pg_dump](https://davejansen.com/how-to-dump-and-restore-a-postgresql-database-from-a-docker-container/) to copy the Postgres DB data with [psql](https://davejansen.com/how-to-dump-and-restore-a-postgresql-database-from-a-docker-container/) to restore.
2) I copied over the Redis `dump.rdb`

However, after starting up a new wger instance, none of my users or data is loaded. I don't have problems backing up and restoring my other docker-compose services. What could I be doing wrong? Has anyone else had success backing up and restoring wger?

What data is saved in Postgres vs. Redis, e.g., user accounts?

**Comment by rolandgeider** [comment]
Hi! The database is definitely the most important thing. If you have uploaded your own exercise images, you will also need to backup the media volume, otherwise it will be recreated the next time you run `manage.py sync-exercises`. Static is 100% generated from the different js and css files on startup. Redis only has some cache so you don't need to to back it.

How are you restoring the db? We're not doing any magic here

**Comment by zero-thermo** [comment]
Hello,

I use this command to backup my main wger instance:

    docker exec -i wger /bin/bash -c "PGPASSWORD=wger pg_dump --username wger wger" > /opt/docker/wger/db-dump/wger_postgres.sql

Then use this command to restore in a separate test instance:

    docker exec -i test-wger /bin/bash -c "PGPASSWORD=test-wger psql --username test-wger test-wger" < /opt/docker/wger/db-dump/wger_postgres.sql

I don't have any uploaded images, so don't need media, static, or redis. I've checked prod.env of my test wger instance and everything there looks correctly updated as well.

Restoring backups for other Postgres DB services works fine with the same command, so I'm stumped as to what I'm doing wrong for wger.

Thanks for having a look.

**Reply by rolandgeider** [reply]
yeah, that looks fine and is similar to what I do. Do you get some errors during startup or something for the web or db services?

**Reply by zero-thermo** [reply]
Startup logs look normal. 

When I run the psql import commands, I see a bunch of errors like this in the db logs:

```
[65] ERROR:  relation "auth_user" does not exist at character 35
[65] STATEMENT:  SELECT COUNT(*) AS "__count" FROM "auth_user"
[126] ERROR:  relation "actstream_action" already exists
[126] STATEMENT:  CREATE TABLE public.actstream_action (
	    id integer NOT NULL,
	    actor_object_id character varying(255) NOT NULL,
	    verb character varying(255) NOT NULL,
	    description text,
	    target_object_id character varying(255),
	    action_object_object_id character varying(255),
	    "timestamp" timestamp with time zone NOT NULL,
	    public boolean NOT NULL,
	    data jsonb,
	    action_object_content_type_id integer,
	    actor_content_type_id integer NOT NULL,
	    target_content_type_id integer
	);
[126] ERROR:  relation "actstream_action_id_seq" already exists
[126] STATEMENT:  CREATE SEQUENCE public.actstream_action_id_seq
	    AS integer
	    START WITH 1
	    INCREMENT BY 1
	    NO MINVALUE
	    NO MAXVALUE
	    CACHE 1;
[126] ERROR:  relation "actstream_follow" already exists
[126] STATEMENT:  CREATE TABLE public.actstream_follow (
	    id integer NOT NULL,
	    object_id character varying(255) NOT NULL,
	    actor_only boolean NOT NULL,
	    started timestamp with time zone NOT NULL,
	    content_type_id integer NOT NULL,
	    user_id integer NOT NULL,
	    flag character varying(255) NOT NULL
	);
```

```
[126] ERROR:  duplicate key value violates unique constraint "auth_group_pkey"
[126] DETAIL:  Key (id)=(1) already exists.
[126] CONTEXT:  COPY auth_group, line 1
[126] STATEMENT:  COPY public.auth_group (id, name) FROM stdin;
[126] ERROR:  duplicate key value violates unique constraint "auth_group_permissions_pkey"
[126] DETAIL:  Key (id)=(1) already exists.
[126] CONTEXT:  COPY auth_group_permissions, line 1
[126] STATEMENT:  COPY public.auth_group_permissions (id, group_id, permission_id) FROM stdin;
```

There are many more varieties of these "already exists" and "duplicate key value" errors. The above is just a small sample.

**Reply by zero-thermo** [reply]
Solved.

Added `--clean` to `pg_dump` to remove all of the "already exists" errors on restore. See [stackoverflow](https://stackoverflow.com/questions/22287914/pg-dump-and-pg-restore-drop-if-exists)
